RFC: Fix skipchars inexact error/hanging (#16892)

fcard opened the pull request:
Hey-ho. I completely forgot about this until I decided to check my issued issues.

This fixes the inexact error raised when using the `skipchars` function. It's also faster for inputs larger than a few characters when you don't use the `linecomment` argument, because it doesn't test each character for it.

I had to change a lot since the function was broken even before that error started to show up (hangs on 0.3), but do tell if this is over-complicating the problem, I am terrible at telling what the line for that is.

Questions:
- This is no longer exclusive to `IOStream`s (works for e.g. `IOBuffer`), should I move it somewhere else? (where?) (actually it doesn't work for `Pipe`s, so maybe I should use a `Union{IOStream,IOBuffer}`?)
- [`peekchar`](https://github.com/JuliaLang/julia/blob/b728145b713e07b60854ad6b81da62c04804a83e/base/iostream.jl#L313), an unexported function only used by the current `skipchars`, is broken. (and I don't know any good way to fix it [I don't think it ever worked as intended, actually]) <s>Should I remove it?</s> I removed it, but please tell me if I should reintroduce it and try to fix it (in this or in another PR).

fcard committed:
Fix skipchars inexact error (#16892)

tkelman commented on base/iostream.jl:
```diff
@@ -321,15 +321,30 @@ function peek(s::IOStream)
+            seek(io,position(io)-sizeof(string(c)))
+            break
+        end
+     end
```
indent slightly off

tkelman commented on test/iostream.jl:
```diff
@@ -0,0 +1,19 @@
+@test eof(skipchars(IOBuffer("#    \n   "), isspace, linecomment='#'))
+
+macro test_skipchars(str, expected_char, lnc=Char(0xffffffff))
+  quote
```
4 space

fcard replied:
Ah gawd, I knew I should have used a automated checker or something. I remember that last time I made a PR incorrect whitespace made the tests fail (and I tested [here](https://github.com/fcard/julia/pull/5) (with a failure on OSX 'coz I am poor)), was that disabled at some point?

Either way I will fix ASAP. Does [ci skip] still work?

tkelman replied:
There's still a failure for trailing whitespace, though we should probably migrate that to a removal script that we just run once a week or so. We don't have an auto indentation checker that I know of, I just have an unusually sensitive eye for it.

[ci skip] does still work, but it's the weekend so things are quiet, no need for it at the moment.

fcard replied:
[[ Had an appointment ]]
Done, thanks for catching these! I will squash it once all the work is finished.

fcard committed:
fix indent ((squash))

fcard committed:
add parentheses to test_skipchars macrocalls ((squash))

The whitespace separating strings containing whitespace was hurting my eyes.

fcard committed:
Remove now unused unexported function peekchar

And its friend, the _chtmp const.

fcard committed:
Make Char(0xffffffff) as a possible value for the linecomment argument ((squash))

fcard committed:
change `action` argument name to `skipcomment` ((squash))

fcard commented:
All right, I think I am done.
Travis tests didn't start for some reason but I got the linux ones here:
* https://travis-ci.org/fcard/julia/jobs/167815584 (i686)
* https://travis-ci.org/fcard/julia/jobs/167815585 (x86_64)

Once this is considered merge-ready, I will squash the commits.
Review away! (no hurry ;P)

tkelman commented:
bumping travis

fcard commented:
I updated my questions. I suddenly won the free-time lottery* (woo country-wide school-blocking protests) so if anybody got any suggestions/requests, this is a good time!

I know this is far from blocking and the julia community isn't going to crumble because of a lack of a `skipchars` function, but I'd like to see this in `0.6` anyway, if for no other reason than that this is a exported function that has been broken since `0.3` and that's no way for a nice language like Julia to behave :P (I hope I don't sound pushy, I know you guys are very busy people and are doing great work, but it would be very nice if someone could take a look at this within the next few weeks)

<sub>*that sounds like I just got fired, but I haven't even finished college yet so there you go</sub>

fcard commented:
Inspired by the latest simplification of my `all`/`any` code, I decided to take a second look at this and through the mystical power of "trusting the compiler" I cut both the size and the complexity in half, without losing any of the performance I gained with manual optimization. (e.g. two closures to avoid an `if` when I could've just trusted Julia to not compare two values if it knows they are of different types)

I think I finally "get" it. When the great river is blocked by the dam, It's not about digging a new way to let the water go around it, it's about poking holes in the dam to let the great river do most of the work itself. 
\*Zen hand gesture\*
<sub>trips in robes and falls over</sub>

StefanKarpinski commented:
What's the status with this?

fcard commented:
@StefanKarpinski I only have those two questions at the top, otherwise I am done with it. I will resolve the conflicts and squash things once I get the a-ok.

tkelman commented:
In what way is peekchar broken? PhyloNetworks and possibly other packages are using it, so can it be deprecated?

fcard commented:
@tkelman Fails on `eof`
```julia
julia> mktemp() do _, file
         Base.peekchar(file)
       end
ERROR: InexactError()
 in convert at ./char.jl:4 [inlined]
 in Type at ./sysimg.jl:66 [inlined]
 in peekchar(::IOStream) at ./iostream.jl:306
 in mktemp(::##7#8, ::String) at ./file.jl:370
 in mktemp(::Function) at ./file.jl:368
```
Because of a `Char(-1)`, which could be substituted by `typemax(Char)` or something I suppose. (which would fix `peekchar` and half the issue with `skipchars` but the latter would still then hang)

fcard commented:
A simpler patch would be:
```diff
function peekchar(s::IOStream)
    if ccall(:ios_peekutf8, Cint, (Ptr{Void}, Ptr{Char}), s, _chtmp) < 0
-        return Char(-1)
+        return typemax(Char)
    end
    return _chtmp[1]
end
```
```diff
-function skipchars(s::IOStream, pred; linecomment::Char=Char(0xffffffff))
+function skipchars(s::IOStream, pred; linecomment=nothing)
    ch = peekchar(s); status = Int(ch)
    while status >= 0 && (pred(ch) || ch == linecomment)
        if ch == linecomment
            readline(s)
        else
            read(s, Char)  # advance one character
         end
        ch = peekchar(s); status = Int(ch)
     end
    return s
end
```
I like my version because it's faster, removes a dependency to a `c` function making this pure julia and this time is an actual simplification, a first for me! But maybe it's better to take the simpler route this time and maybe later when things are calmer taking a second look at this. The release date is pretty close isn't it?

Also this version doesn't work with `IOBuffer`s so the tests are going to need to be a little bit more elaborate :P

fcard committed:
Simplify function

fcard committed:
Restore peekchar

fcard commented:
Piss and balls! I am never trusting that "Resolve Conflicts" button again. I just clicked on it, changed the one required line, pressed the shiny buttons and look at the mess it made :( <sub>why do they always have to merge and never rebase...</sub>
(also my linux blew up a while ago and that messed with my file permissions somehow, but whatever)

Anyway, I'd like to know the veredito on this. I will just follow instructions, I don't want any say on this.
(I fixed and restored the `peekchar` function either way)

fcard committed:
Mfing whitespace

fcard commented:
Hi, I know you're busy but I need to cut this off as soon as possible. Please either reject it or detail the steps I need to take to have this merged.

tkelman commented:
This seems reasonable, though we're in feature freeze which is not an ideal time for IO changes. We could always revert and revisit after branching if it happens to cause any problems. Do you think this version would have compatibility consequences for the packagws that use this?

fcard commented:
I don't think so, this version is a superset of the old one in terms of functionality.

However, I can make the [simpler change](https://github.com/JuliaLang/julia/pull/18752#issuecomment-270259034) for now and revisit this later, if you want to be absolutely safe. Packages would only change behavior if they are depending in the bug in some way like

```julia
try
  skipchars(func,file)
catch err
  if err == InexactError
     println("Why would you do this?")
  end
end
```

tkelman commented:
That's pretty clearly buggy. Is the widening of `skipchars` to general `IO` more a feature or a bugfix?

fcard commented:
Sorry, power outage.

The support for IOBuffer/etc was a convenient side-effect of the new algorithm, a "why not" kinda deal. It made testing easier too. I can remove that for now. (I should be more strict than `IO` anyway since it doesn't work with Pipes)

fcard committed:
Remove support for IOBuffer

fcard committed:
Trailling whitespace + typo

fcard commented:
I don't think that 32bit failure on windows is related, is it?

I spent the day reviewing the code again since it's been a while, and I didn't realize `seek` was so slow for `IOStream`s! The new version is still 2x faster for larger inputs but has 10x overhead, because of that [`seek`](https://github.com/fcard/julia/blob/2e51efcbe50a811998f1093d07479e6ce93cc997/base/iostream.jl#L330) :(

I solved this by writing a new c+julia function `seekback` that just moves the position back n bytes, and a `_charbytes` function since that `sizeof(string(c))` method wasn't doing too hot either, so the function is just strictly better now (hopefully). That's over [>>>here<<<](https://github.com/fcard/julia/pull/7). I think that's too many last minute changes with things I don't have much experience with.

Sorry I took so long.

tkelman commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/18752#issuecomment-284104604) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/2b02591cc8807cc5bd9b6a6b62053c1180a76976/6ba8805_vs_1c90ed6/report.md). cc @jrevels

fcard commented:
Oh. Well... That's done then. I started drawing something to represent my feelings towards the eventual conclusion of this...

![the_grateful_death_of_blanc_an_art_piece1](https://cloud.githubusercontent.com/assets/12222927/23583354/fa464bdc-0120-11e7-98c7-bc8193f7c9a7.png)
But it looked like an indie art piece where they died and the soul went off there on top right saying "Phew, finally off the confines of the mortal world". (the simple casual onomatopoeia "Phew" in contrast with the grim situation is `Pathos`, you see. Ducks)
So I tried wiping up a tiny anime to better represent my feelings:
![blanc_phew_1](https://cloud.githubusercontent.com/assets/12222927/23583373/6eb27e1e-0121-11e7-88e3-2d6efaaf97e5.png)
Doesn't look that good because my linux fricked and I <sub>had to install mint and the drivers for the wacom tablet <sub> were outdated and I don't want to use packages anymore and now I need to compile <sub> llvm but it doesn't find libffi even through I have it installed and...</sub></sub></sub>

---

(god I love that `<sub>` tag so much, I hope it doesn't annoy anybody. You can always safely ignore anything in tiny letters :P)

ANYWAY. Sorry for hurrying this up so close to the `0.6` release but unresolved things leech away at my brain like nothing else, and I need to focus on some important things right now. Next time I make a contribution I will consult the community to see if there is enough interest and not too much going on so the PR can be resolved more quickly, or if it's wanted at all.

---

**If there is any performance problems** consider looking at the changes I made in my `aaaaaa` commit over [here](https://github.com/fcard/julia/pull/7/commits/bfc5a5ffb0a5de804102b1425f7cf1dac27208de) (The tests there are better too if you want to use them).

Thank you again for your patience and I will go back hiding in my blankets (or going back to college or whatever)

Love y'all.

StefanKarpinski commented:
I have to confess I've completely lost track of what the story is here...

1. I think there's a performance problem
2. But there's also a potential widening of functionality
3. It may be backwards compatible (i.e. the functionality expansion wouldn't break code)

We can always fix 1 alone any time, regardless of feature freezes and whatnot. If 3 is true, then we could even do 2 after a feature freeze. If 3 is false, then it would be great to at least get the part of the change that fixes 1 and sets the stage for 2 in the future.

fcard commented:
Hmm, summarizing:

  - I modified a function, `skipchars`, making it:
    - not have the bug anymore 
    - faster for large inputs
    - also working with `IOBuffers`
       - (not a goal initially but was just too easy a change to make in the end)

  - Since we're on feature freeze, I removed the IOBuffer part, so it's just bugfree and faster now.

  - I realized that my function has some performance regression on some types of input, so I tried to
    make a fix for that. I managed it, but it was not trivial so I did not include it in this PR. The changes
    are on the [`aaaa` commit](https://github.com/fcard/julia/pull/7/commits/bfc5a5ffb0a5de804102b1425f7cf1dac27208de), if anybody wants to take a look and bring them over.

  - This got merged. and I ((<img src="https://cloud.githubusercontent.com/assets/12222927/23583373/6eb27e1e-0121-11e7-88e3-2d6efaaf97e5.png" width=64, height=64 alt="badly drawn sigh"></img>)) But to be completely I don't know if this was completed resolved or not, I was just trusting the higher power here.

StefanKarpinski commented:
I don't think that broadening the applicability to IOBuffers is a problem post-feature freeze – after all, it won't break anyone's code, and avoiding breakage of already-ported code is the purpose of feature freeze. So we could reintroduce that broadening, IMO. @tkelman tends to have stricter views on these things and may disagree. Fixing additional performance regression before 0.6 final would be great, but technically, it can happen any time in the 0.6 cycle.
